File: app/src/main/java/org/tasks/jobs/MidnightRefreshWork.java
package org.tasks.jobs;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.work.WorkerParameters;
import javax.inject.Inject;
import org.tasks.LocalBroadcastManager;
import org.tasks.injection.JobComponent;

public class MidnightRefreshWork extends RepeatingWorker {

  @Inject WorkManager workManager;
  @Inject LocalBroadcastManager localBroadcastManager;

  public MidnightRefreshWork(@NonNull Context context, @NonNull WorkerParameters workerParams) {
    super(context, workerParams);
  }

  @Override
  protected Result run() {
    localBroadcastManager.broadcastRefresh();
    return Result.success();
  }

  @Override
  protected void scheduleNext() {
    workManager.scheduleMidnightRefresh();
  }

  @Override
  protected void inject(JobComponent component) {
    component.inject(this);
  }
}

File: app/src/main/java/org/tasks/jobs/MidnightRefreshWork.kt
package org.tasks.jobs

import android.content.Context
import androidx.hilt.work.HiltWorker
import androidx.work.WorkerParameters
import dagger.assisted.Assisted
import dagger.assisted.AssistedInject
import org.tasks.LocalBroadcastManager
import org.tasks.analytics.Firebase

@HiltWorker
class MidnightRefreshWork @AssistedInject constructor(
        @Assisted context: Context,
        @Assisted workerParams: WorkerParameters,
        firebase: Firebase,
        private val workManager: WorkManager,
        private val localBroadcastManager: LocalBroadcastManager) : RepeatingWorker(context, workerParams, firebase) {

    override suspend fun run(): Result {
        localBroadcastManager.broadcastRefresh()
        return Result.success()
    }

    override suspend fun scheduleNext() = workManager.scheduleMidnightRefresh()
}